File: services/assessment-service/src/test/java/com/visor/school/assessment/model/AssessmentTest.java
package com.visor.school.assessment.model;

import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;

import static org.junit.jupiter.api.Assertions.*;

class AssessmentTest {

    @Test
    void shouldCreateAssessmentWithRequiredFields() {
        Assessment assessment = new Assessment(
            UUID.randomUUID(),
            "Midterm Exam",
            AssessmentType.EXAM,
            new BigDecimal("100.0"),
            UUID.randomUUID()
        );

        assertNotNull(assessment.getId());
        assertEquals("Midterm Exam", assessment.getName());
        assertEquals(AssessmentType.EXAM, assessment.getType());
        assertEquals(0, new BigDecimal("100.0").compareTo(assessment.getTotalPoints()));
        assertEquals(AssessmentStatus.DRAFT, assessment.getStatus());
        assertNull(assessment.getDescription());
        assertNull(assessment.getWeight());
        assertNull(assessment.getDueDate());
    }

    @Test
    void shouldCreateAssessmentWithAllFields() {
        UUID classId = UUID.randomUUID();
        UUID teacherId = UUID.randomUUID();
        LocalDate dueDate = LocalDate.now().plusDays(30);

        Assessment assessment = new Assessment(
            classId,
            "Final Project",
            AssessmentType.PROJECT,
            new BigDecimal("150.0"),
            teacherId,
            "Comprehensive project",
            new BigDecimal("30.0"),
            dueDate
        );

        assertEquals(classId, assessment.getClassId());
        assertEquals("Final Project", assessment.getName());
        assertEquals(AssessmentType.PROJECT, assessment.getType());
        assertEquals("Comprehensive project", assessment.getDescription());
        assertEquals(0, new BigDecimal("150.0").compareTo(assessment.getTotalPoints()));
        assertEquals(0, new BigDecimal("30.0").compareTo(assessment.getWeight()));
        assertEquals(dueDate, assessment.getDueDate());
        assertEquals(teacherId, assessment.getCreatedBy());
        assertEquals(AssessmentStatus.DRAFT, assessment.getStatus());
    }

    @Test
    void shouldValidateTotalPointsIsPositive() {
        assertThrows(IllegalArgumentException.class, () -> {
            new Assessment(
                UUID.randomUUID(),
                "Test",
                AssessmentType.TEST,
                new BigDecimal("-10.0"),
                UUID.randomUUID()
            );
        });
    }

    @Test
    void shouldValidateWeightIsBetween0And100() {
        assertThrows(IllegalArgumentException.class, () -> {
            new Assessment(
                UUID.randomUUID(),
                "Test",
                AssessmentType.TEST,
                new BigDecimal("100.0"),
                UUID.randomUUID(),
                null,
                new BigDecimal("150.0"),
                null
            );
        });
    }

    @Test
    void shouldPublishAssessment() throws InterruptedException {
        Assessment assessment = new Assessment(
            UUID.randomUUID(),
            "Quiz 1",
            AssessmentType.QUIZ,
            new BigDecimal("50.0"),
            UUID.randomUUID()
        );

        var initialUpdatedAt = assessment.getUpdatedAt();
        Thread.sleep(10);

        assessment.publish();

        assertEquals(AssessmentStatus.PUBLISHED, assessment.getStatus());
        assertTrue(assessment.getUpdatedAt().isAfter(initialUpdatedAt));
    }

    @Test
    void shouldMarkAssessmentAsGrading() {
        Assessment assessment = new Assessment(
            UUID.randomUUID(),
            "Test",
            AssessmentType.TEST,
            new BigDecimal("100.0"),
            UUID.randomUUID()
        );
        assessment.publish();

        assessment.markAsGrading();

        assertEquals(AssessmentStatus.GRADING, assessment.getStatus());
    }

    @Test
    void shouldCompleteAssessment() {
        Assessment assessment = new Assessment(
            UUID.randomUUID(),
            "Test",
            AssessmentType.TEST,
            new BigDecimal("100.0"),
            UUID.randomUUID()
        );
        assessment.publish();
        assessment.markAsGrading();

        assessment.complete();

        assertEquals(AssessmentStatus.COMPLETED, assessment.getStatus());
    }

    @Test
    void shouldAcceptAllAssessmentTypes() {
        List<AssessmentType> types = Arrays.asList(
            AssessmentType.TEST,
            AssessmentType.QUIZ,
            AssessmentType.ASSIGNMENT,
            AssessmentType.PROJECT,
            AssessmentType.EXAM,
            AssessmentType.FINAL_EXAM
        );

        for (AssessmentType type : types) {
            Assessment assessment = new Assessment(
                UUID.randomUUID(),
                "Test " + type,
                type,
                new BigDecimal("100.0"),
                UUID.randomUUID()
            );

            assertEquals(type, assessment.getType());
        }
    }
}

File: services/academic-assessment-service/src/test/kotlin/com/visor/school/assessment/model/AssessmentTest.kt
package com.visor.school.assessment.model

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.math.BigDecimal
import java.time.LocalDate
import java.time.LocalDateTime
import java.util.UUID

class AssessmentTest {

    @Test
    fun `should create assessment with required fields`() {
        val assessment = Assessment(
            classId = UUID.randomUUID(),
            name = "Midterm Exam",
            type = AssessmentType.EXAM,
            totalPoints = BigDecimal("100.0"),
            createdBy = UUID.randomUUID()
        )

        assertNotNull(assessment.id)
        assertEquals("Midterm Exam", assessment.name)
        assertEquals(AssessmentType.EXAM, assessment.type)
        assertEquals(BigDecimal("100.0"), assessment.totalPoints)
        assertEquals(AssessmentStatus.DRAFT, assessment.status)
        assertNull(assessment.description)
        assertNull(assessment.weight)
        assertNull(assessment.dueDate)
    }

    @Test
    fun `should create assessment with all fields`() {
        val classId = UUID.randomUUID()
        val teacherId = UUID.randomUUID()
        val dueDate = LocalDate.of(2024, 12, 15)

        val assessment = Assessment(
            classId = classId,
            name = "Final Project",
            type = AssessmentType.PROJECT,
            description = "Comprehensive project",
            totalPoints = BigDecimal("150.0"),
            weight = BigDecimal("30.0"),
            dueDate = dueDate,
            createdBy = teacherId
        )

        assertEquals(classId, assessment.classId)
        assertEquals("Final Project", assessment.name)
        assertEquals(AssessmentType.PROJECT, assessment.type)
        assertEquals("Comprehensive project", assessment.description)
        assertEquals(BigDecimal("150.0"), assessment.totalPoints)
        assertEquals(BigDecimal("30.0"), assessment.weight)
        assertEquals(dueDate, assessment.dueDate)
        assertEquals(teacherId, assessment.createdBy)
        assertEquals(AssessmentStatus.DRAFT, assessment.status)
    }

    @Test
    fun `should validate total points is positive`() {
        assertThrows<IllegalArgumentException> {
            Assessment(
                classId = UUID.randomUUID(),
                name = "Test",
                type = AssessmentType.TEST,
                totalPoints = BigDecimal("-10.0"),
                createdBy = UUID.randomUUID()
            )
        }
    }

    @Test
    fun `should validate weight is between 0 and 100`() {
        assertThrows<IllegalArgumentException> {
            Assessment(
                classId = UUID.randomUUID(),
                name = "Test",
                type = AssessmentType.TEST,
                totalPoints = BigDecimal("100.0"),
                weight = BigDecimal("150.0"),
                createdBy = UUID.randomUUID()
            )
        }
    }

    @Test
    fun `should publish assessment`() {
        val assessment = Assessment(
            classId = UUID.randomUUID(),
            name = "Quiz 1",
            type = AssessmentType.QUIZ,
            totalPoints = BigDecimal("50.0"),
            createdBy = UUID.randomUUID()
        )

        val initialUpdatedAt = assessment.updatedAt
        Thread.sleep(10)

        assessment.publish()

        assertEquals(AssessmentStatus.PUBLISHED, assessment.status)
        assertTrue(assessment.updatedAt.isAfter(initialUpdatedAt))
    }

    @Test
    fun `should mark assessment as grading`() {
        val assessment = Assessment(
            classId = UUID.randomUUID(),
            name = "Test",
            type = AssessmentType.TEST,
            totalPoints = BigDecimal("100.0"),
            createdBy = UUID.randomUUID()
        )
        assessment.publish()

        assessment.markAsGrading()

        assertEquals(AssessmentStatus.GRADING, assessment.status)
    }

    @Test
    fun `should complete assessment`() {
        val assessment = Assessment(
            classId = UUID.randomUUID(),
            name = "Test",
            type = AssessmentType.TEST,
            totalPoints = BigDecimal("100.0"),
            createdBy = UUID.randomUUID()
        )
        assessment.publish()
        assessment.markAsGrading()

        assessment.complete()

        assertEquals(AssessmentStatus.COMPLETED, assessment.status)
    }

    @Test
    fun `should accept all assessment types`() {
        val types = listOf(
            AssessmentType.TEST,
            AssessmentType.QUIZ,
            AssessmentType.ASSIGNMENT,
            AssessmentType.PROJECT,
            AssessmentType.EXAM,
            AssessmentType.FINAL_EXAM
        )

        types.forEach { type ->
            val assessment = Assessment(
                classId = UUID.randomUUID(),
                name = "Test $type",
                type = type,
                totalPoints = BigDecimal("100.0"),
                createdBy = UUID.randomUUID()
            )

            assertEquals(type, assessment.type)
        }
    }
}

